### DISCUSSION TITLE
Saturation "energy" in the Gaussian gain model

### DISCUSSION THREAD

**Discussion by skynet-programmer69** [AaHaHaa/MMTools#12] [Q&A] [answered]
Hi Yi-Hao,

First I would like to say you are the GOAT for releasing & maintaining this code + gas_UPPE as open source repositories. I'm a big fan of your work and the unique theories/capabilities you and the Wise group have produced, so thank you for the work you have done.

I've only just begun to learn how this code base works and how to modify it for my own goals, but while looking into the implemented Gaussian gain model I noticed the definition of saturation_energy (explicitly discussed in the function GMMNLSE_propagate_with_adaptive, for instance) actually gives the units of saturation power: saturation_energy = saturation_intensity * Aeff = h * f/(sigma * tau) * Aeff -> m^2*J/(s*m^2) -> W. 

This becomes an issue in line 74 of the function "stepping_RK4IP_SMGaussianGain_adaptive" as the exponential dispersion + gain factor is: expDG = exp((D_op + G/(1+total_energy/saturation_energy))*sim.dz/2); where total_energy is calculated with the nested function calc_total_energy and gives total_energy in nJ. With these definitions the exponential will no longer be left dimensionless and can obviously give rogue results.

It may be that I missed something along the chain of functions that completely negates everything I claimed above, but either way I'd rather be safe than sorry.

Any elucidation on this potential problem would be much appreciated.

Again thanks for all the work you have done and for releasing this for all to enjoy!

**Comment by AaHaHaa** [comment, marked as answer]
Oh, thanks for finding this.

It's the mistake with the comments. The code is still correct.

It needs to be multiplied by t_rep, the roundtrip time or the inverse of the pulse repetition rate.
In the steady-state gain assumption, the gain medium cannot see each pulse but an overall optical effect/saturation; equivalently, this means that the gain medium sees only the smeared energy over one period of pulsed laser operation. This is why the saturation should be multiplied by t_rep, making it into energy unit to be dividable by the pulse energy.

Smear the pulse: saturation term = P_pulse/P_sat = (E_pulse/t_rep)/P_sat = E_pulse/(P_sat * t_rep) = E_pulse/E_sat, so E_sat = P_sat * t_rep.

I updated the code with updated comments.

It's all my honor that you can use this code, which is the purpose I share it out.
And many thanks for helping me make it better by fixing things that I miss myself.

**Reply by skynet-programmer69** [reply]
This makes sense, bit of confusion from me and the code comment. Glad it was nothing serious.

Thanks for the prompt reply!
